Remote peers on 0.0.0.0 instead of 127.0.0.1

In running the tests, I'm getting failures due to expecting 127.0.0.1 but getting back 0.0.0.0:
```pytb
________________________________ test_tcp_repr _________________________________

    def test_func():
        with pristine_loop() as loop:
            if iscoroutinefunction(func):
                cor = func
            else:
                cor = gen.coroutine(func)
            try:
>               loop.run_sync(cor, timeout=timeout)

distributed/utils_test.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:581: in run_sync
    return future_cell[0].result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:1113: in run
    yielded = self.gen.send(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @gen_test()
    def test_tcp_repr():
        a, b = yield get_tcp_comm_pair()
        assert a.local_address in repr(b)
>       assert b.local_address in repr(a)
E       AssertionError: assert 'tcp://127.0.0.1:39127' in '<TCP local=tcp://127.0.0.1:33746 remote=tcp://0.0.0.0:39127>'
E        +  where 'tcp://127.0.0.1:39127' = <TCP local=tcp://127.0.0.1:39127 remote=tcp://127.0.0.1:33746>.local_address
E        +  and   '<TCP local=tcp://127.0.0.1:33746 remote=tcp://0.0.0.0:39127>' = repr(<TCP local=tcp://127.0.0.1:33746 remote=tcp://0.0.0.0:39127>)

distributed/comm/tests/test_comms.py:1085: AssertionError
```
```pytb
________________________________ test_tls_repr _________________________________

    def test_func():
        with pristine_loop() as loop:
            if iscoroutinefunction(func):
                cor = func
            else:
                cor = gen.coroutine(func)
            try:
>               loop.run_sync(cor, timeout=timeout)

distributed/utils_test.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:581: in run_sync
    return future_cell[0].result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:1113: in run
    yielded = self.gen.send(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @gen_test()
    def test_tls_repr():
        a, b = yield get_tls_comm_pair()
        assert a.local_address in repr(b)
>       assert b.local_address in repr(a)
E       AssertionError: assert 'tls://127.0.0.1:34149' in '<TLS local=tls://127.0.0.1:51272 remote=tls://0.0.0.0:34149>'
E        +  where 'tls://127.0.0.1:34149' = <TLS local=tls://127.0.0.1:34149 remote=tls://127.0.0.1:51272>.local_address
E        +  and   '<TLS local=tls://127.0.0.1:51272 remote=tls://0.0.0.0:34149>' = repr(<TLS local=tls://127.0.0.1:51272 remote=tls://0.0.0.0:34149>)

distributed/comm/tests/test_comms.py:1093: AssertionError
--------------------------- Captured stderr teardown ---------------------------
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:33746 remote=tcp://0.0.0.0:39127>
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:39127 remote=tcp://127.0.0.1:33746>
```
```pytb
______________________________ test_tcp_adresses _______________________________

    def test_func():
        with pristine_loop() as loop:
            if iscoroutinefunction(func):
                cor = func
            else:
                cor = gen.coroutine(func)
            try:
>               loop.run_sync(cor, timeout=timeout)

distributed/utils_test.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:581: in run_sync
    return future_cell[0].result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:1107: in run
    yielded = self.gen.throw(*exc_info)
distributed/comm/tests/test_comms.py:1116: in test_tcp_adresses
    yield check_addresses(a, b)
/usr/lib64/python3.7/site-packages/tornado/gen.py:1099: in run
    value = future.result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:296: in wrapper
    result = func(*args, **kwargs)
/usr/lib64/python3.7/types.py:277: in wrapped
    coro = func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = <TCP local=tcp://127.0.0.1:49700 remote=tcp://0.0.0.0:34603>
b = <TCP local=tcp://127.0.0.1:34603 remote=tcp://127.0.0.1:49700>

    @gen.coroutine
    def check_addresses(a, b):
>       assert a.peer_address == b.local_address
E       AssertionError: assert 'tcp://0.0.0.0:34603' == 'tcp://127.0.0.1:34603'
E         - tcp://0.0.0.0:34603
E         ?       ^     ^
E         + tcp://127.0.0.1:34603
E         ?       ^^^     ^

distributed/comm/tests/test_comms.py:1107: AssertionError
```
```pytb
______________________________ test_tls_adresses _______________________________

    def test_func():
        with pristine_loop() as loop:
            if iscoroutinefunction(func):
                cor = func
            else:
                cor = gen.coroutine(func)
            try:
>               loop.run_sync(cor, timeout=timeout)

distributed/utils_test.py:769: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:581: in run_sync
    return future_cell[0].result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:1107: in run
    yielded = self.gen.throw(*exc_info)
distributed/comm/tests/test_comms.py:1122: in test_tls_adresses
    yield check_addresses(a, b)
/usr/lib64/python3.7/site-packages/tornado/gen.py:1099: in run
    value = future.result()
/usr/lib64/python3.7/site-packages/tornado/gen.py:296: in wrapper
    result = func(*args, **kwargs)
/usr/lib64/python3.7/types.py:277: in wrapped
    coro = func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = <TLS local=tls://127.0.0.1:39508 remote=tls://0.0.0.0:40249>
b = <TLS local=tls://127.0.0.1:40249 remote=tls://127.0.0.1:39508>

    @gen.coroutine
    def check_addresses(a, b):
>       assert a.peer_address == b.local_address
E       AssertionError: assert 'tls://0.0.0.0:40249' == 'tls://127.0.0.1:40249'
E         - tls://0.0.0.0:40249
E         ?       ^     ^
E         + tls://127.0.0.1:40249
E         ?       ^^^     ^

distributed/comm/tests/test_comms.py:1107: AssertionError
```

I also see a failure to _not_ connect in `test_hostport` and `test_server_listen`, which may be because something's listening on all addresses instead of a specific one.

I think this one might related too, as it seems like it might be listening on 0.0.0.0 when it shouldn't:
```pytb
______________________________ test_server_listen ______________________________

    def test_func():
        with pristine_loop() as loop:
            if iscoroutinefunction(func):
                cor = func
            else:
                cor = gen.coroutine(func)
            try:
>               loop.run_sync(cor, timeout=timeout)

distributed/utils_test.py:813: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:571: in run_sync
    self.start()
/usr/lib64/python3.7/site-packages/tornado/platform/asyncio.py:132: in start
    self.asyncio_loop.run_forever()
/usr/lib64/python3.7/asyncio/base_events.py:539: in run_forever
    self._run_once()
/usr/lib64/python3.7/asyncio/base_events.py:1775: in _run_once
    handle._run()
/usr/lib64/python3.7/asyncio/events.py:88: in _run
    self._context.run(self._callback, *self._args)
/usr/lib64/python3.7/site-packages/tornado/ioloop.py:758: in _run_callback
    ret = callback()
/usr/lib64/python3.7/site-packages/tornado/stack_context.py:300: in null_wrapper
    return fn(*args, **kwargs)
/usr/lib64/python3.7/site-packages/tornado/gen.py:1233: in inner
    self.run()
/usr/lib64/python3.7/site-packages/tornado/gen.py:1147: in run
    yielded = self.gen.send(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

addr = 'tcp://0.0.0.0:7811', timeout = 0.5, connection_args = None
exception_class = <class 'OSError'>

    @gen.coroutine
    def assert_cannot_connect(
        addr, timeout=None, connection_args=None, exception_class=EnvironmentError
    ):
        """
        Check that it is impossible to connect to the distributed *addr*
        within the given *timeout*.
        """
        if timeout is None:
            timeout = 0.5
        with pytest.raises(exception_class):
            comm = yield connect(addr, timeout=timeout, connection_args=connection_args)
>           comm.abort()
E           Failed: DID NOT RAISE <class 'OSError'>

distributed/utils_test.py:1219: Failed
----------------------------- Captured stderr call -----------------------------
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP  local=tcp://127.0.0.1:47078 remote=tcp://0.0.0.0:35831>
```

--- a/distributed/utils.py
+++ b/distributed/utils.py
@@ -170,12 +170,20 @@ def get_fileno_limit():
 
 @toolz.memoize
 def _get_ip(host, port, family):
+    def hostname_fallback():
+        addr_info = socket.getaddrinfo(
+            socket.gethostname(), port, family, socket.SOCK_DGRAM, socket.IPPROTO_UDP
+        )[0]
+        return addr_info[4][0]
+
     # By using a UDP socket, we don't actually try to connect but
     # simply select the local address through which *host* is reachable.
     sock = socket.socket(family, socket.SOCK_DGRAM)
     try:
         sock.connect((host, port))
         ip = sock.getsockname()[0]
+        if ip == "0.0.0.0":
+            return hostname_fallback()
         return ip
     except OSError as e:
         warnings.warn(
@@ -183,10 +191,7 @@ def _get_ip(host, port, family):
             "reaching %r, defaulting to hostname: %s" % (host, e),
             RuntimeWarning,
         )
-        addr_info = socket.getaddrinfo(
-            socket.gethostname(), port, family, socket.SOCK_DGRAM, socket.IPPROTO_UDP
-        )[0]
-        return addr_info[4][0]
+        return hostname_fallback()
     finally:
         sock.close()
 

